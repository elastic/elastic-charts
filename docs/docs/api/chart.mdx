---
id: Chart
sidebar_position: 0
sidebar_label: Chart
---

# Top-level `Chart` Component

## Overview {#overview}

The whole `@elastic/charts` library starts with the `Chart` component.

## Props {#props}

### `size`
- Type: <code>size?: [ChartSize](/all-types/type-aliases/ChartSize)</code>
- Default: `'100%'`

Sets the size of the chart.

Examples:
- `200` - Sets width and height to 200px
- `'90%'` - Sets width and height to 90% of parent
- `['100%', 200]` - Sets width to 100% of parent and height to 200px
- `{ width: '100%', height: 200 }` - Sets width to 100% of parent and height to 200px

### `id`
- Type: `id?: string`
- Default: a random UUID via [`uuid`](https://www.npmjs.com/package/uuid)

A unique id to indetify the chart. Used to renerate individual chart state.

### `title`
- Type: `title?: string`

Top-level title of the chart.

### `description`
- Type: `description?: string`

Top-level description of the chart.

### `children`
- Type: `children?: ReactNode`

The children **specs** defining the chart.

In `@elastic/charts` we configure the chart using a react component api, we call these configuration components **specs**. Theses specs are applied as `children` of the `Chart` element.

While technically `children` are optional, passing no children will simply return an empty chart.


### `className`
- Type: `className?: string`

Adds className to `.echChartContent` element for custom styling.

:::warning

Careful adding too many styles, as this can lead to unexpected apperance and behaviour of the chart.

:::

### `renderer`
- Type: `renderer?: 'svg' | 'canvas'`
- Default: `'canvas'`

This sets the renderer type of the chart.

:::info
Currently this option not configurable per chart as there is not a parrallel renderer for all charts.
:::
