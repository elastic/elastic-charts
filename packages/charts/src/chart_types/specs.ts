/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

export {
  AreaSeries,
  AreaSeriesProps,
  Axis,
  AxisProps,
  BarSeries,
  BarSeriesProps,
  BubbleSeries,
  BubbleSeriesProps,
  HistogramBarSeries,
  HistogramBarSeriesProps,
  LineAnnotation,
  LineAnnotationProps,
  LineSeries,
  LineSeriesProps,
  RectAnnotation,
  RectAnnotationProps,
} from './xy_chart/specs';

export * from './xy_chart/utils/specs';

export * from './wordcloud/specs';

export * from './goal_chart/specs';

export { Partition } from './partition_chart/specs';

export { Heatmap, HeatmapSpec, RasterTimeScale, TimeScale, LinearScale, OrdinalScale } from './heatmap/specs';

export {
  Metric,
  MetricSpecProps,
  MetricSpec,
  MetricBase,
  MetricWText,
  MetricWNumber,
  MetricWNumberArrayValues,
  MetricWStringArrayValues,
  MetricWProgress,
  MetricWTrend,
  MetricTrendShape,
  MetricDatum,
  getMetricValue,
} from './metric/specs';

export { Bullet, BulletProps, BulletSpec, BulletDatum, BulletSubtype, BulletValueLabels } from './bullet_graph/spec';
export { BulletStyle } from './bullet_graph/theme';
