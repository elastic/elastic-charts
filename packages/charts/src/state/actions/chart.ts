/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

/** @internal */
export const CHART_RENDERED = 'CHART_RENDERED';

interface ChartRenderedAction {
  type: typeof CHART_RENDERED;
}

/** @internal */
export function onChartRendered(): ChartRenderedAction {
  return { type: CHART_RENDERED };
}

/** @internal */
export type ChartActions = ChartRenderedAction;
