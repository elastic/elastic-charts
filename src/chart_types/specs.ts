export {
  AreaSeries,
  BarSeries,
  LineSeries,
  Axis,
  LineAnnotation,
  RectAnnotation,
  HistogramBarSeries,
} from './xy_chart/specs';
