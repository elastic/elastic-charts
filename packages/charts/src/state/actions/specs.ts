/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { Spec } from '../../specs';

/** @internal */
export const UPSERT_SPEC = 'UPSERT_SPEC';

/** @internal */
export const REMOVE_SPEC = 'REMOVE_SPEC';

/** @internal */
export const SPEC_PARSED = 'SPEC_PARSED';

/** @internal */
export const SPEC_UNMOUNTED = 'SPEC_UNMOUNTED';

interface SpecParsedAction {
  type: typeof SPEC_PARSED;
}

interface SpecUnmountedAction {
  type: typeof SPEC_UNMOUNTED;
}

interface UpsertSpecAction {
  type: typeof UPSERT_SPEC;
  spec: Spec;
}

interface RemoveSpecAction {
  type: typeof REMOVE_SPEC;
  id: string;
}

/** @internal */
export function upsertSpec(spec: Spec): UpsertSpecAction {
  return { type: UPSERT_SPEC, spec };
}

/** @internal */
export function removeSpec(id: string): RemoveSpecAction {
  return { type: REMOVE_SPEC, id };
}

/** @internal */
export function specParsed(): SpecParsedAction {
  return { type: SPEC_PARSED };
}

/** @internal */
export function specUnmounted(): SpecUnmountedAction {
  return { type: SPEC_UNMOUNTED };
}

/** @internal */
export type SpecActions = SpecParsedAction | SpecUnmountedAction | UpsertSpecAction | RemoveSpecAction;
