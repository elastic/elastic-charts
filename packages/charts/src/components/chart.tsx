/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import classNames from 'classnames';
import React, { createRef } from 'react';
import { Provider } from 'react-redux';
import { createStore, Store, Unsubscribe, StoreEnhancer, applyMiddleware, Middleware } from 'redux';
import { v4 as uuidv4 } from 'uuid';

import { Colors } from '../common/colors';
import { LegendPositionConfig, PointerEvent } from '../specs';
import { SpecsParser } from '../specs/specs_parser';
import { onExternalPointerEvent } from '../state/actions/events';
import { onComputedZIndex } from '../state/actions/z_index';
import { chartStoreReducer, GlobalChartState } from '../state/chart_state';
import { getInternalIsInitializedSelector, InitStatus } from '../state/selectors/get_internal_is_intialized';
import { getLegendConfigSelector } from '../state/selectors/get_legend_config_selector';
import { ChartSize, getChartSize } from '../utils/chart_size';
import { LayoutDirection } from '../utils/common';
import { ChartBackground } from './chart_background';
import { ChartContainer } from './chart_container';
import { ChartResizer } from './chart_resizer';
import { ChartStatus } from './chart_status';
import { ErrorBoundary } from './error_boundary';
import { Legend } from './legend/legend';
import { getElementZIndex } from './portal/utils';

interface ChartProps {
  /**
   * The type of rendered
   * @defaultValue `canvas`
   */
  renderer?: 'svg' | 'canvas';
  size?: ChartSize;
  className?: string;
  id?: string;
}

interface ChartState {
  legendDirection: LegendPositionConfig['direction'];
}

const getMiddlware = (id: string): StoreEnhancer => {
  const middlware: Middleware<any, any, any>[] = [];

  // eslint-disable-next-line no-underscore-dangle
  if (typeof window !== 'undefined' && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-call, no-underscore-dangle
    return (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
      trace: true,
      name: `@elastic/charts (id: ${id})`,
    })(applyMiddleware(...middlware));
  }

  return applyMiddleware(...middlware);
};

/** @public */
export class Chart extends React.Component<ChartProps, ChartState> {
  static defaultProps: ChartProps = {
    renderer: 'canvas',
  };

  private unsubscribeToStore: Unsubscribe;

  private chartStore: Store<GlobalChartState>;

  private chartContainerRef: React.RefObject<HTMLDivElement>;

  private chartStageRef: React.RefObject<HTMLCanvasElement>;

  constructor(props: ChartProps) {
    super(props);
    this.chartContainerRef = createRef();
    this.chartStageRef = createRef();

    const id = props.id ?? uuidv4();
    const storeReducer = chartStoreReducer(id);
    const enhancer = getMiddlware(id);
    this.chartStore = createStore(storeReducer, enhancer);
    this.state = {
      legendDirection: LayoutDirection.Vertical,
    };
    this.unsubscribeToStore = this.chartStore.subscribe(() => {
      const state = this.chartStore.getState();
      if (getInternalIsInitializedSelector(state) !== InitStatus.Initialized) {
        return;
      }

      const {
        legendPosition: { direction },
      } = getLegendConfigSelector(state);
      if (this.state.legendDirection !== direction) {
        this.setState({
          legendDirection: direction,
        });
      }
      if (state.internalChartState) {
        state.internalChartState.eventCallbacks(state);
      }
    });
  }

  componentDidMount() {
    if (this.chartContainerRef.current) {
      const zIndex = getElementZIndex(this.chartContainerRef.current, document.body);
      this.chartStore.dispatch(onComputedZIndex(zIndex));
    }
  }

  componentWillUnmount() {
    this.unsubscribeToStore();
  }

  getPNGSnapshot(
    // eslint-disable-next-line unicorn/no-object-as-default-parameter
    options = {
      backgroundColor: Colors.Transparent.keyword,
    },
  ): {
    blobOrDataUrl: any;
    browser: 'IE11' | 'other';
  } | null {
    if (!this.chartStageRef.current) {
      return null;
    }
    const canvas = this.chartStageRef.current;
    const backgroundCanvas = document.createElement('canvas');
    backgroundCanvas.width = canvas.width;
    backgroundCanvas.height = canvas.height;
    const bgCtx = backgroundCanvas.getContext('2d');
    if (!bgCtx) {
      return null;
    }
    bgCtx.fillStyle = options.backgroundColor;
    bgCtx.fillRect(0, 0, canvas.width, canvas.height);
    bgCtx.drawImage(canvas, 0, 0);

    return {
      blobOrDataUrl: backgroundCanvas.toDataURL(),
      browser: 'other',
    };
  }

  getChartContainerRef = () => this.chartContainerRef;

  dispatchExternalPointerEvent(event: PointerEvent) {
    this.chartStore.dispatch(onExternalPointerEvent(event));
  }

  render() {
    const { size, className } = this.props;
    const containerSizeStyle = getChartSize(size);
    const chartClassNames = classNames('echChart', className, {
      'echChart--column': this.state.legendDirection === LayoutDirection.Horizontal,
    });

    return (
      <Provider store={this.chartStore}>
        <div className={chartClassNames} style={containerSizeStyle}>
          <ChartBackground />
          <ChartStatus />
          <ChartResizer />
          <Legend />
          {/* TODO: Add renderFn to error boundary */}
          <ErrorBoundary>
            <SpecsParser>{this.props.children}</SpecsParser>
            <div className="echContainer" ref={this.chartContainerRef}>
              <ChartContainer getChartContainerRef={this.getChartContainerRef} forwardStageRef={this.chartStageRef} />
            </div>
          </ErrorBoundary>
        </div>
      </Provider>
    );
  }
}
