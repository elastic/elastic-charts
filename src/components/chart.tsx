import React, { CSSProperties, createRef } from 'react';
import classNames from 'classnames';
import { Provider } from 'react-redux';
import { createStore, Store } from 'redux';
import uuid from 'uuid';
import { SpecsParser } from '../specs/specs_parser';
import { ChartResizer } from './chart_resizer';
import { Legend } from './legend/legend';
import { ChartContainer } from './chart_container';
import { isHorizontalAxis } from '../chart_types/xy_chart/utils/axis_utils';
import { Position } from '../utils/commons';
import { ChartSize, getChartSize } from '../utils/chart_size';
import { ChartStatus } from './chart_status';
import { chartStoreReducer, GlobalChartState } from '../state/chart_state';
import { isInitialized } from '../state/selectors/is_initialized';
import { createOnElementOutCaller } from '../chart_types/xy_chart/state/selectors/on_element_out_caller';
import { createOnElementOverCaller } from '../chart_types/xy_chart/state/selectors/on_element_over_caller';
import { createOnElementClickCaller } from '../chart_types/xy_chart/state/selectors/on_element_click_caller';
import { ChartTypes } from '../chart_types/index';
import { getSettingsSpecSelector } from '../state/selectors/get_settings_specs';
import { createOnBrushEndCaller } from '../chart_types/xy_chart/state/selectors/on_brush_end_caller';
import { onExternalPointerEvent } from '../state/actions/events';
import { PointerEvent } from '../specs';
import { createOnPointerMoveCaller } from '../chart_types/xy_chart/state/selectors/on_pointer_move_caller';

interface ChartProps {
  /** The type of rendered
   * @default 'canvas'
   */
  renderer: 'svg' | 'canvas';
  size?: ChartSize;
  className?: string;
  id?: string;
}

interface ChartState {
  legendPosition: Position;
}

function getContainerStyle(size: any): CSSProperties {
  if (size) {
    return {
      position: 'relative',
      ...getChartSize(size),
    };
  }
  return {};
}

export class Chart extends React.Component<ChartProps, ChartState> {
  static defaultProps: ChartProps = {
    renderer: 'canvas',
  };
  private chartStore: Store<GlobalChartState>;
  private chartContainerRef: React.RefObject<HTMLDivElement>;
  private chartStageRef: React.RefObject<HTMLCanvasElement>;

  constructor(props: any) {
    super(props);
    this.chartContainerRef = createRef();
    this.chartStageRef = createRef();

    const id = uuid.v4();
    const storeReducer = chartStoreReducer(id);
    const enhancers =
      typeof window !== 'undefined' && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
        ? (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ trace: true, name: `@elastic/charts (id: ${id})` })()
        : undefined;

    this.chartStore = createStore(storeReducer, enhancers);
    this.state = {
      legendPosition: Position.Right,
    };

    const onElementClickCaller = createOnElementClickCaller();
    const onElementOverCaller = createOnElementOverCaller();
    const onElementOutCaller = createOnElementOutCaller();
    const onBrushEndCaller = createOnBrushEndCaller();
    const onPointerMoveCaller = createOnPointerMoveCaller();
    this.chartStore.subscribe(() => {
      const state = this.chartStore.getState();
      if (!isInitialized(state)) {
        return;
      }
      const settings = getSettingsSpecSelector(state);
      if (this.state.legendPosition !== settings.legendPosition) {
        this.setState({
          legendPosition: settings.legendPosition,
        });
      }

      if (state.chartType !== ChartTypes.XYAxis) {
        return;
      }
      onElementOverCaller(state);
      onElementOutCaller(state);
      onElementClickCaller(state);
      onBrushEndCaller(state);
      onPointerMoveCaller(state);
    });
  }

  dispatchExternalPointerEvent(event: PointerEvent) {
    this.chartStore.dispatch(onExternalPointerEvent(event));
  }

  getPNGSnapshot(
    options = {
      backgroundColor: 'transparent',
      pixelRatio: 2,
    },
  ): {
    blobOrDataUrl: any;
    browser: 'IE11' | 'other';
  } | null {
    if (!this.chartStageRef.current) {
      return null;
    }
    const canvas = this.chartStageRef.current;
    const backgroundCanvas = document.createElement('canvas');
    backgroundCanvas.width = canvas.width;
    backgroundCanvas.height = canvas.height;
    const bgCtx = backgroundCanvas.getContext('2d');
    if (!bgCtx) {
      return null;
    }
    bgCtx.fillStyle = options.backgroundColor;
    bgCtx.fillRect(0, 0, canvas.width, canvas.height);
    bgCtx.drawImage(canvas, 0, 0);

    // @ts-ignore
    if (bgCtx.msToBlob) {
      // @ts-ignore
      const blobOrDataUrl = bgCtx.msToBlob();
      return {
        blobOrDataUrl,
        browser: 'IE11',
      };
    } else {
      return {
        blobOrDataUrl: backgroundCanvas.toDataURL(),
        browser: 'other',
      };
    }
  }

  getChartContainerRef = () => {
    return this.chartContainerRef;
  };

  render() {
    const { size, className } = this.props;
    const containerStyle = getContainerStyle(size);
    const horizontal = isHorizontalAxis(this.state.legendPosition);
    const chartClassNames = classNames('echChart', className, {
      'echChart--column': horizontal,
    });
    return (
      <Provider store={this.chartStore}>
        <div style={containerStyle} className={chartClassNames} ref={this.chartContainerRef}>
          <ChartStatus />
          <ChartResizer />
          <Legend />
          <SpecsParser>{this.props.children}</SpecsParser>
          <div className="echContainer">
            <ChartContainer getChartContainerRef={this.getChartContainerRef} forwardStageRef={this.chartStageRef} />
          </div>
        </div>
      </Provider>
    );
  }
}
