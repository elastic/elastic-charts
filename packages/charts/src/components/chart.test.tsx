/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { mount } from 'enzyme';
import React from 'react';

import { Chart } from './chart';
import { Settings, BarSeries } from '../specs';

interface Datum {
  x: number;
  y: number;
}

describe('Chart', () => {
  it("should render 'No data to display' without series", () => {
    const wrapper = mount(<Chart />);
    expect(wrapper.text()).toContain('No data to display');
  });

  it("should render 'No data to display' with settings but without series", () => {
    const wrapper = mount(
      <Chart>
        <Settings debug rendering="svg" />
      </Chart>,
    );
    expect(wrapper.text()).toContain('No data to display');
  });

  it("should render 'No data to display' with an empty dataset", () => {
    const wrapper = mount(
      <Chart size={[100, 100]}>
        <Settings debug rendering="svg" />
        <BarSeries<Datum> id="test" data={[]} xAccessor="x" yAccessors={['y']} />
      </Chart>,
    );
    expect(wrapper.text()).toContain('No data to display');
  });

  it('should render the legend name test', () => {
    // This is snapshot has a slight diff between local and ci runs, need to investigate
    // See https://github.com/elastic/elastic-charts/runs/4706561347?check_suite_focus=true
    const wrapper = mount(
      <Chart size={[100, 100]} id="chart1">
        <Settings debug rendering="svg" showLegend />
        <BarSeries id="test" data={[{ x: 0, y: 2 }]} xAccessor="x" yAccessors={['y']} />
      </Chart>,
    );
    expect(wrapper.debug()).toMatchSnapshot();
  });
});
