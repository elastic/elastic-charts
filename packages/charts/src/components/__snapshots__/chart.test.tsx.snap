// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart should render the legend name test 1`] = `
"<Chart size={{...}} id=\\"chart1\\" renderer=\\"canvas\\">
  <Provider store={{...}}>
    <div className=\\"echChart\\" style={{...}}>
      <Connect(ChartBackground)>
        <ChartBackground backgroundColor=\\"transparent\\" dispatch={[Function: dispatch]}>
          <div className=\\"echChartBackground\\" style={{...}} />
        </ChartBackground>
      </Connect(ChartBackground)>
      <Connect(ChartStatusComponent)>
        <ChartStatusComponent chartId=\\"chart1\\" rendered={true} renderedCount={1} onRenderChange={[undefined]} debugState={{...}} dispatch={[Function: dispatch]}>
          <div className=\\"echChartStatus\\" data-ech-render-complete={true} data-ech-render-count={1} data-ech-debug-state={{...}} />
        </ChartStatusComponent>
      </Connect(ChartStatusComponent)>
      <Connect(Resizer)>
        <Resizer resizeDebounce={10} updateParentDimensions={[Function (anonymous)]}>
          <div className=\\"echChartResizer\\" />
        </Resizer>
      </Connect(Resizer)>
      <Connect(LegendComponent)>
        <LegendComponent debug={true} chartDimensions={{...}} containerDimensions={{...}} chartTheme={{...}} size={{...}} items={{...}} extraValues={{...}} config={{...}} onToggleDeselectSeriesAction={[Function (anonymous)]} onItemOutAction={[Function (anonymous)]} onItemOverAction={[Function (anonymous)]} clearTemporaryColors={[Function (anonymous)]} setTemporaryColor={[Function (anonymous)]} setPersistedColor={[Function (anonymous)]}>
          <div className=\\"echLegend echLegend--debug echLegend--vertical echLegend--right echLegend--top\\" style={{...}}>
            <div style={{...}} className=\\"echLegendListContainer\\">
              <ul style={{...}} className=\\"echLegendList\\">
                <LegendItem item={{...}} positionConfig={{...}} totalItems={1} extraValues={{...}} showExtra={false} onMouseOut={[undefined]} onMouseOver={[undefined]} onClick={[undefined]} clearTemporaryColorsAction={[Function (anonymous)]} setPersistedColorAction={[Function (anonymous)]} setTemporaryColorAction={[Function (anonymous)]} mouseOutAction={[Function (anonymous)]} mouseOverAction={[Function (anonymous)]} toggleDeselectSeriesAction={[Function (anonymous)]} colorPicker={[undefined]} action={[undefined]}>
                  <li className=\\"echLegendItem echLegendItem--vertical\\" onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} style={[undefined]} data-ech-series-name=\\"test\\">
                    <div className=\\"background\\" />
                    <ForwardRef color=\\"#1EA593\\" seriesName=\\"test\\" isSeriesHidden={false} hasColorPicker={false} onClick={[undefined]}>
                      <div className=\\"echLegendItem__color\\" title=\\"series color\\">
                        <Icon type=\\"dot\\" color=\\"#1EA593\\" aria-label=\\"series color: #1EA593\\">
                          <DotIcon className=\\"echIcon\\" color=\\"#1EA593\\" tabIndex={[undefined]} focusable=\\"false\\" aria-label=\\"series color: #1EA593\\">
                            <svg width={16} height={16} viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\" className=\\"echIcon\\" color=\\"#1EA593\\" tabIndex={[undefined]} focusable=\\"false\\" aria-label=\\"series color: #1EA593\\">
                              <circle cx={8} cy={8} r={4} />
                            </svg>
                          </DotIcon>
                        </Icon>
                      </div>
                    </ForwardRef>
                    <Label label=\\"test\\" isToggleable={false} onClick={[undefined]} isSeriesHidden={false}>
                      <div className=\\"echLegendItem__label\\" title=\\"test\\" onClick={[undefined]}>
                        test
                      </div>
                    </Label>
                  </li>
                </LegendItem>
              </ul>
            </div>
          </div>
        </LegendComponent>
      </Connect(LegendComponent)>
      <ErrorBoundary>
        <Connect(SpecsParserComponent)>
          <SpecsParserComponent specParsed={[Function (anonymous)]} specUnmounted={[Function (anonymous)]}>
            <Connect(SpecInstance) debug={true} rendering=\\"svg\\" showLegend={true}>
              <SpecInstance debug={true} rendering=\\"svg\\" showLegend={true} upsertSpec={[Function (anonymous)]} removeSpec={[Function (anonymous)]} id=\\"__global__settings___\\" chartType=\\"global\\" specType=\\"settings\\" rotation={0} animateData={true} resizeDebounce={10} tooltip={{...}} pointerUpdateTrigger=\\"x\\" externalPointerEvents={{...}} hideDuplicateAxes={false} baseTheme={{...}} brushAxis=\\"x\\" minBrushDelta={2} ariaUseDefaultSummary={true} ariaLabelHeadingLevel=\\"p\\" showLegendExtra={false} legendMaxDepth={Infinity} legendPosition=\\"right\\" />
            </Connect(SpecInstance)>
            <Connect(SpecInstance) id=\\"test\\" data={{...}}>
              <SpecInstance id=\\"test\\" data={{...}} upsertSpec={[Function (anonymous)]} removeSpec={[Function (anonymous)]} chartType=\\"xy_axis\\" specType=\\"series\\" seriesType=\\"bar\\" groupId=\\"__global__\\" xScaleType=\\"ordinal\\" yScaleType=\\"linear\\" xAccessor=\\"x\\" yAccessors={{...}} hideInLegend={false} enableHistogramMode={false} />
            </Connect(SpecInstance)>
          </SpecsParserComponent>
        </Connect(SpecsParserComponent)>
        <div className=\\"echContainer\\">
          <Connect(ChartContainer) getChartContainerRef={[Function (anonymous)]} forwardStageRef={{...}}>
            <ChartContainer getChartContainerRef={[Function (anonymous)]} forwardStageRef={{...}} status=\\"Initialized\\" initialized={true} isChartEmpty={false} pointerCursor=\\"default\\" isBrushingAvailable={false} isBrushing={false} internalChartRenderer={[Function (anonymous)]} settings={{...}} onPointerMove={[Function (anonymous)]} onMouseUp={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onKeyPress={[Function (anonymous)]}>
              <div className=\\"echChartPointerContainer\\" style={{...}} onMouseMove={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onMouseUp={[Function (anonymous)]}>
                <Connect(Crosshair)>
                  <Crosshair theme={{...}} chartRotation={0} cursorPosition={[undefined]} cursorCrossLinePosition={[undefined]} tooltipType=\\"vertical\\" fromExternalEvent={[undefined]} zIndex={0} dispatch={[Function: dispatch]} />
                </Connect(Crosshair)>
                <Connect(XYChart) forwardCanvasRef={{...}}>
                  <XYChart forwardCanvasRef={{...}} initialized={true} isChartEmpty={false} debug={true} geometries={{...}} geometriesIndex={{...}} theme={{...}} chartContainerDimensions={{...}} highlightedLegendItem={[undefined]} rotation={0} renderingArea={{...}} chartTransform={{...}} axesSpecs={{...}} perPanelAxisGeoms={{...}} perPanelGridLines={{...}} axesStyles={{...}} annotationDimensions={{...}} annotationSpecs={{...}} panelGeoms={{...}} a11ySettings={{...}} onChartRendered={[Function (anonymous)]}>
                    <figure aria-labelledby={[undefined]} aria-describedby=\\"chart1--defaultSummary\\">
                      <canvas className=\\"echCanvasRenderer\\" width={150} height={200} style={{...}} role=\\"presentation\\" />
                    </figure>
                    <Connect(ScreenReaderSummaryComponent)>
                      <ScreenReaderSummaryComponent chartTypeDescription=\\"bar chart\\" a11ySettings={{...}} goalChartData={{...}} goalChartLabels={{...}} dispatch={[Function: dispatch]}>
                        <div className=\\"echScreenReaderOnly\\">
                          <ScreenReaderLabel label={[undefined]} labelId={[undefined]} labelHeadingLevel=\\"p\\" description={[undefined]} descriptionId=\\"chart1--defaultSummary\\" defaultSummaryId=\\"chart1--defaultSummary\\" tableCaption={[undefined]} goalChartLabels={{...}} />
                          <ScreenReaderDescription label={[undefined]} labelId={[undefined]} labelHeadingLevel=\\"p\\" description={[undefined]} descriptionId=\\"chart1--defaultSummary\\" defaultSummaryId=\\"chart1--defaultSummary\\" tableCaption={[undefined]} />
                          <ScreenReaderTypes label={[undefined]} labelId={[undefined]} labelHeadingLevel=\\"p\\" description={[undefined]} descriptionId=\\"chart1--defaultSummary\\" defaultSummaryId=\\"chart1--defaultSummary\\" tableCaption={[undefined]} chartTypeDescription=\\"bar chart\\" goalChartData={{...}}>
                            <dl>
                              <dt>
                                Chart type:
                              </dt>
                              <dd id=\\"chart1--defaultSummary\\">
                                bar chart
                              </dd>
                            </dl>
                          </ScreenReaderTypes>
                        </div>
                      </ScreenReaderSummaryComponent>
                    </Connect(ScreenReaderSummaryComponent)>
                  </XYChart>
                </Connect(XYChart)>
                <Connect(Tooltip) getChartContainerRef={[Function (anonymous)]}>
                  <Tooltip getChartContainerRef={[Function (anonymous)]} visible={false} zIndex={0} info={{...}} position={{...}} headerFormatter={[undefined]} settings={{...}} rotation={0} chartId=\\"chart1\\" backgroundColor=\\"transparent\\" onPointerMove={[Function (anonymous)]} />
                </Connect(Tooltip)>
                <Connect(Annotations) getChartContainerRef={[Function (anonymous)]} chartAreaRef={{...}}>
                  <Annotations getChartContainerRef={[Function (anonymous)]} chartAreaRef={{...}} isChartEmpty={false} chartDimensions={{...}} annotationDimensions={{...}} annotationSpecs={{...}} tooltipState={{...}} chartId=\\"chart1\\" zIndex={0} onPointerMove={[Function (anonymous)]} onDOMElementLeave={[Function (anonymous)]} onDOMElementEnter={[Function (anonymous)]}>
                    <AnnotationTooltip chartId=\\"chart1\\" zIndex={0} state={{...}} chartRef={{...}} onScroll={[Function (anonymous)]} />
                  </Annotations>
                </Connect(Annotations)>
                <Connect(Highlighter)>
                  <Highlighter initialized={true} chartId=\\"chart1\\" zIndex={0} highlightedGeometries={{...}} chartTransform={{...}} chartDimensions={{...}} chartRotation={0} dispatch={[Function: dispatch]}>
                    <svg className=\\"echHighlighter\\" style={{...}}>
                      <defs>
                        <clipPath id=\\"echHighlighterClipPath__chart1\\">
                          <rect x=\\"0\\" y=\\"0\\" width={130} height={180} />
                        </clipPath>
                      </defs>
                    </svg>
                  </Highlighter>
                </Connect(Highlighter)>
                <Connect(BrushTool)>
                  <BrushTool initialized={true} projectionContainer={{...}} mainProjectionArea={{...}} isBrushAvailable={false} isBrushing={false} brushArea={{...}} zIndex={0} dispatch={[Function: dispatch]} />
                </Connect(BrushTool)>
              </div>
            </ChartContainer>
          </Connect(ChartContainer)>
        </div>
      </ErrorBoundary>
    </div>
  </Provider>
</Chart>"
`;
